runtime/compiler/control/JitDump.cpp

```diff
@@ -526,174 +517,198 @@ runJitdump(char *label, J9RASdumpContext *context, J9RASdumpAgent *agent)
-               recompilationThreadInfo->resumeCompilationThread();  // TODO: Postpone this so that the thread does not get to sleep again
-               if (options->getVerboseOption(TR_VerboseDump))
-                  TR_VerboseLog::writeLineLocked(TR_Vlog_JITDUMP, "Resuming DiagCompThread");
+            if (options->getVerboseOption(TR_VerboseDump))
+               TR_VerboseLog::writeLineLocked(TR_Vlog_JITDUMP, "Found compilation object");
 
-               // get old start PC if method was available
-               void *oldStartPC = 0;
-               if (currentMethodBeingCompiled)
-                  oldStartPC = currentMethodBeingCompiled->_oldStartPC;
+            // dump IL of current compilation
+            dumpCurrentIL(comp, crashedThread, jitConfig, logFile);
 
-               // request the compilation
-               TR_CompilationErrorCode compErrCode;
-               compErrCode = recompileMethodForLog(
-                  crashedThread,
-                  (J9Method *)(comp->getCurrentMethod()->getPersistentIdentifier()),
-                  compInfo,
-                  frontendOfThread,
-                  (TR_Hotness)comp->getOptLevel(),
-                  comp->isProfilingCompilation(),
-                  oldStartPC,
-                  logFile
+            // if there was an available compilation thread, recompile the current method
+            if (recompilationThread)
+               {
+               // only proceed to recompile if the method is a regular Java method
+               if (currentMethodBeingCompiled &&
+                  currentMethodBeingCompiled->getMethodDetails().isOrdinaryMethod())
+                  {
+                  // resume the healthy compilation thread
+                  recompilationThreadInfo->resumeCompilationThread();  // TODO: Postpone this so that the thread does not get to sleep again
+                  if (options->getVerboseOption(TR_VerboseDump))
+                     TR_VerboseLog::writeLineLocked(TR_Vlog_JITDUMP, "Resuming DiagCompThread");
+
+                  // get old start PC if method was available
+                  void *oldStartPC = 0;
+                  if (currentMethodBeingCompiled)
+                     oldStartPC = currentMethodBeingCompiled->_oldStartPC;
+
+                  // The current thread is a compilation thread which may or may not currently hold VM access. The request
+                  // for recompilation to generate the jitdump will be performed synchronously for which the code path we
+                  // will take will be the same as if we were an application thread. We will take the synchronous request
+                  // path in `compileOnSeparateThread` which ends up releasing VM access prior to waiting on the diagnostic
+                  // thread to finish the queued compile. To avoid deadlocks we must acquire VM access here.
+                  TR::VMAccessCriticalSection requestSynchronousCompilation(TR_J9VMBase::get(jitConfig, crashedThread));
```

Comment:
The `releaseVMAccess` macro that we call in `compileOnSeparateThread` should have asserted I would have thought, because prior to this change we would not have held VM access while trying to release it. Any ideas why we don't assert in today's world?

Reply:
It looks like in [1] we go to call [2]:

```
VM_VMAccess::inlineReleaseVMAccess (vmThread=0x8c200) at /openjdk/openj9/runtime/oti/VMAccess.hpp:179
```

There is no assert there. This looks like the "fast" version of release VM access. Looks dangerous IMO. Should we be calling the API which does assert throughout the JIT?

[1] https://github.com/eclipse/openj9/blob/3fe51ff39281bf1c5e316b021164eea41256febe/runtime/compiler/control/CompilationThread.cpp#L6223
[2] https://github.com/eclipse/openj9/blob/3fe51ff39281bf1c5e316b021164eea41256febe/runtime/oti/VMAccess.hpp#L165-L193